### PR TITLE
Support signed S3 download URLs

### PR DESCRIPTION
**Pull request by crux-capacitor**
> PRs that do not follow the template will be automatically closed

## References

- [ ] I have updated the [documentation](https://github.com/sorry-cypress/gitbook). PR link `<here>`
- [ ] I have added inclued automated tests
- [ ] I acknowledge that I have tested that the change is backwards-compatible
- [ ] Original issue / feature request / discussion: `<here>`

## Use case

> For security reasons, the S3 screenshots bucket in my deployment was required to be encrypted with a CMK. This started leading to 400 Bad Request responses from S3 when the browser would try to download the video file for a test.

To get around this, SC should return signed download URLs.

## Example

> Submit screenshots / outputs / tests together with the PR.

### CODE REVIEW

File: cloudformation/sorry-cypress.yml [hunk 1]
```diff
@@ -313,7 +313,6 @@ Resources:
     Type: AWS::S3::Bucket
     Properties:
       BucketName: !Join ['-', [!Ref 'AWS::StackName', 'test-results']]
-      AccessControl: PublicRead
```
Comment (crux-capacitor) [inline cloudformation/sorry-cypress.yml:316]: This resolves a breaking change due to AWS updating allowed default settings on new buckets.

File: packages/dashboard/src/instance/instanceDetails.tsx [hunk 2]
```diff
@@ -359,6 +374,13 @@ function getTreeOfNavigationItems(instance: GetInstanceQuery['instance']) {
 
   // if instance has video add it as the first item of the navigation tree
   if (instance?.results?.videoUrl) {
+
+    if (S3_PRESIGN_DOWNLOAD_URLS === true) {
+      const key = instance.results.videoUrl.split('/').pop()  // split url and get last item to get object key
+      const signedDownloadUrl = s3.getDownloadUrl(key)
+      instance.results.videoUrl = signedDownloadUrl
```
Comment (crux-capacitor) [inline packages/dashboard/src/instance/instanceDetails.tsx:381]: This `split` doesn't work - it says `Property 'split' does not exist on type 'string'.`. How can I make this work?

File: packages/director/src/screenshots/s3/config.ts [hunk 3]
```diff
@@ -5,7 +5,12 @@ export const S3_FORCE_PATH_STYLE = process.env.S3_FORCE_PATH_STYLE || false;
 export const S3_READ_URL_PREFIX = process.env.S3_READ_URL_PREFIX || undefined;
 export const S3_IMAGE_KEY_PREFIX = process.env.S3_IMAGE_KEY_PREFIX || undefined;
 export const S3_VIDEO_KEY_PREFIX = process.env.S3_VIDEO_KEY_PREFIX || undefined;
-export const UPLOAD_EXPIRY_SECONDS = parseInt(
-  process.env.UPLOAD_EXPIRY_SECONDS || '90',
+export const S3_PRESIGN_DOWNLOAD_URLS = process.env.S3_PRESIGN_DOWNLOAD_URLS || false;
```
Comment (crux-capacitor) [inline packages/director/src/screenshots/s3/config.ts:8]: This is the new bool variable that triggers generating signed download URLs.

File: cloudformation/sorry-cypress.yml (hunk 1 above)
Comment (agoldis) [reply cloudformation/sorry-cypress.yml:316]: Sorry for my ignorance, what do you mean here?

File: packages/dashboard/src/instance/instanceDetails.tsx [hunk 4]
```diff
@@ -3,6 +3,21 @@ import ChevronRightIcon from '@mui/icons-material/ChevronRight';
...
+  S3_ACL,
+  S3_BUCKET,
+  S3_REGION,
+  S3_PRESIGN_DOWNLOAD_URLS,
+  S3_FORCE_PATH_STYLE,
+  S3_READ_URL_PREFIX,
+  S3_IMAGE_KEY_PREFIX,
+  S3_VIDEO_KEY_PREFIX,
+  S3_DOWNLOAD_EXPIRY_SECONDS,
+} from '../../../director/src/screenshots/s3/config';
```
Comment (agoldis) [inline packages/dashboard/src/instance/instanceDetails.tsx:18]: I am not sure you can use the config values (in runtime) from other package. One approach would be to generate signed URLs on the API side

File: packages/director/src/screenshots/s3/config.ts (hunk 3 above)
Comment (agoldis) [reply packages/director/src/screenshots/s3/config.ts:8]: The previous variable `UPLOAD_EXPIRY_SECONDS` was being used in minio driver as well, we need to preserve it

File: packages/director/src/screenshots/s3/s3.ts [hunk 5]
```diff
@@ -74,5 +82,47 @@ export const getVideoUploadUrl = async (
...
+  key,
+  Expires = S3_DOWNLOAD_EXPIRY_SECONDS,
+}: GetDownloadURLParams): Promise<S3SignedDownloadResult> => {
+  const s3Params = {
+    Bucket: S3_BUCKET,
+    Key: key,
+    Expires
+  };
+
+  return new Promise((resolve, reject) => {
```
Comment (agoldis) [inline packages/director/src/screenshots/s3/s3.ts:99]: The signed download URL will be persisted to the DB with the expired signature. I think the signing should be done on the API each time we retrieve the url

**Review by agoldis — CHANGES_REQUESTED** [review]
Seems like it needs more work - please see the inline comments

**Review by agoldis — CHANGES_REQUESTED** [review]
Seems like it needs more work - please see the inline comments

**Comment by agoldis** [comment]
@crux-capacitor

**Pull request closed without merging**